demos/sample_azure_iot/sample_azure_iot.c

```diff
@@ -0,0 +1,625 @@
+                                      uint32_t * pulIothubDeviceIdLength )
+    {
+        NetworkContext_t xNetworkContext = { 0 };
+        TlsTransportParams_t xTlsTransportParams = { 0 };
+        TlsTransportStatus_t xNetworkStatus;
+        AzureIoTProvisioningClientResult_t xResult;
+        AzureIoTTransportInterface_t xTransport;
+        uint32_t ucSamplepIothubHostnameLength = sizeof( ucSampleIotHubHostname );
+        uint32_t ucSamplepIothubDeviceIdLength = sizeof( ucSampleIotHubDeviceId );
+
+        /* Set the pParams member of the network context with desired transport. */
+        xNetworkContext.pParams = &xTlsTransportParams;
+
+        xNetworkStatus = prvConnectToServerWithBackoffRetries( democonfigENDPOINT, democonfigMQTT_BROKER_PORT,
+                                                               pXNetworkCredentials, &xNetworkContext );
+        configASSERT( xNetworkStatus == TLS_TRANSPORT_SUCCESS );
+
+        /* Fill in Transport Interface send and receive function pointers. */
+        xTransport.pxNetworkContext = &xNetworkContext;
+        xTransport.xSend = TLS_Socket_Send;
+        xTransport.xRecv = TLS_Socket_Recv;
+
+        xResult = AzureIoTProvisioningClient_Init( &xAzureIoTProvisioningClient,
+                                                   ( const uint8_t * ) democonfigENDPOINT,
+                                                   sizeof( democonfigENDPOINT ) - 1,
+                                                   ( const uint8_t * ) democonfigID_SCOPE,
+                                                   sizeof( democonfigID_SCOPE ) - 1,
+                                                   ( const uint8_t * ) democonfigREGISTRATION_ID,
+                                                   sizeof( democonfigREGISTRATION_ID ) - 1,
+                                                   NULL, ucSharedBuffer, sizeof( ucSharedBuffer ),
+                                                   ullGetUnixTime,
+                                                   &xTransport );
+        configASSERT( xResult == eAzureIoTProvisioningSuccess );
+
+        #ifdef democonfigDEVICE_SYMMETRIC_KEY
+            xResult = AzureIoTProvisioningClient_SetSymmetricKey( &xAzureIoTProvisioningClient,
+                                                                  ( const uint8_t * ) democonfigDEVICE_SYMMETRIC_KEY,
+                                                                  sizeof( democonfigDEVICE_SYMMETRIC_KEY ) - 1,
+                                                                  Crypto_HMAC );
+            configASSERT( xResult == eAzureIoTProvisioningSuccess );
+        #endif // democonfigDEVICE_SYMMETRIC_KEY
+
+        do
+        {
+            xResult = AzureIoTProvisioningClient_Register( &xAzureIoTProvisioningClient,
+                                                           sampleazureiotProvisioning_Registration_TIMEOUT_MS );
+        } while( xResult == eAzureIoTProvisioningPending );
```

Comment:
Nit: IIRC there's various "limit max number of retries down in the _Register" function.  Or maybe not, if we always get pending??.

In any case I think we sholud have some limit of max number of times we'll go through this loop.  Unfortunately if DPS is down and always pending there's not a ton to be done but it could at least be used a best practice to show a bail-out.  How app would actually handle such a thing - maybe via a watchdog or longer timeout - is way outside scope of this straightforward sample of course.

Reply:
Currently assumption is that state machine in DPS always move to Error state, if any failure happen in any state (like sending or response failure). As MQTT currently sits over TCP and we do MQTT keep alive, we will bail out of this loop as soon as network error/service error occurs. 

Based on above I wanted this logic to be simple. Please let me know if you think otherwise.

Reply:
If the product code does the right thing - and we have it contractually obligated to do so always (which I think back-compat 101 would guarantee?) then agree with keeping sample simple.